### PR TITLE
Tests for inventory and modules

### PR DESCRIPTION
**Pull request by Faustasm**

### CODE REVIEW

File: test/conftest.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        return [
+            FirewallRule(**firewall_rule)
+            for firewall_rule in data['firewall_rules']['firewall_rule']
+        ]
+
+    def read_json_data(self, filename):
+        cwd = os.path.dirname(__file__)
+        with open('{}/json_data/{}.json'.format(cwd, filename), 'r') as json_file:
+            data = json.load(json_file)
+        json_file.close()
```
Comment (abamaxa) [inline test/conftest.py]: I don't think this is needed because the context manager will close the file

File: test/test_firewall.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+from itertools import product
+
+class TestFirewall(object):
+    def test_match_firewall_rules(self, firewall_manager):
+        firewall_rules = firewall_manager.manager.get_firewall_rules()
+        for rule in firewall_rules:
+            firewall_manager.match_firewall_rules(rule, firewall_rules)
```
Comment (abamaxa) [inline test/test_firewall.py]: Can we add some assertions to check that `match_firewall_rules` has behaved as expected.

File: test/test_firewall.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+from itertools import product
+
+class TestFirewall(object):
+    def test_match_firewall_rules(self, firewall_manager):
+        firewall_rules = firewall_manager.manager.get_firewall_rules()
```
Comment (abamaxa) [inline test/test_firewall.py:5]: Please add an assertion to check firewall_rules is not empty

File: test/test_inventory.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
+from itertools import product
+from inventory.upcloud import list_servers, get_server
+
+class TestInventory(object):
+    def test_list_servers(self, manager):
+        IPvs_to_test = ['IPv4', 'IPv6']
+        possible_configs = list(product([True, False], repeat=2))
+
```
Comment (abamaxa) [inline test/test_inventory.py:8]: Can we add a check that `possible_configs` is not empty

File: test/test_inventory.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+        for IP_v in IPvs_to_test:
+            for config in possible_configs:
+                list_servers(manager, config[0], config[1], IP_v)
```
Comment (abamaxa) [inline test/test_inventory.py]: Please add some assertions to check the `list_servers` returns the expected result

File: test/test_inventory.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+    def test_get_server(self, manager):
+        possible_configs = list(product([True, False], repeat=3))
+        for config in possible_configs:
+            get_server(manager, config[0], config[1], config[2])
```
Comment (abamaxa) [inline test/test_inventory.py]: Please add assertions to check if `possible_configs` is empty and that `get_server` behaves as expected

File: test/test_tag.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
+class TestTag(object):
+    def test_create_missing_tags(self, tag_manager):
+        tag_manager.create_missing_tags('test')
+
```
Comment (abamaxa) [inline test/test_tag.py:6]: Please check whether this succeeds

File: test/test_upcloud.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
+from itertools import product
+from inventory.upcloud import list_servers, get_server
+
+class TestUpcloud(object):
+    def test_find_server(self, server_manager):
+        server_uuid = '008c365d-d307-4501-8efc-cd6d3bb0e494'
+        server_hostname = 'fi.example.com'
+        possible_configs = list(product([server_uuid, server_hostname, None], repeat=2))
+
```
Comment (abamaxa) [inline test/test_upcloud.py:10]: Please check that `possible_configs` is not empty

File: tox.ini [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+exclude =
+    # No need to traverse our git directory
+    .git,
+    # There's no value in checking cache directories
+    __pycache__,
+    # ...or build ones
+    build,
+    # Ignore virtualenv
+    ENV
+max-line-length = 99
```
Comment (abamaxa) [inline tox.ini:45]: But otherwise, its a very good start, thanks

**Comment by Faustasm** [comment]
Ok. Will update pr with changes.

**Pull request merged**